Key technical improvements

mdhaber opened the pull request.

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -877,30 +873,28 @@ \subsubsection*{Sparse matrices}
-in a \texttt{scipy.LowLevelCallable()} object that reduces the overhead for
-calling compiled C functions directly from Python.  Supported low-level
-functions include \texttt{PyCapsule} objects, ctypes function pointers, and
-cffi function pointers. The low level function signature must be consistent
```
It's unclear to me whether/why the details about the function signature are important to include in the paper. The bit about "low level function signature must be consistent with the expectations of the routine it is passed to" seems redundant with "as long as the function call signature matches the specifications". I don't think we need to specify that "The C code may be generated using numba or Cython" as there is no reason for the reader to suspect that the source of the C code matters. The `scipy.ndimage.generic\_filter` example doesn't strengthen the text because the reader is probably unfamiliar with `scipy.ndimage.generic\_filter`.
I'd be happy to help with revisions, though, if this part is really important.

tylerjereddy replied:
`numba` and `Cython` are major open source projects that provide useful ways to interface with Python for preformance reasons, so the interop seems relevant to me.

Some cut down is probably fine as you say.

mdhaber replied:
How about 
"As of SciPy version 0.19, it is possible for users to wrap low-level functions
in a \texttt{scipy.LowLevelCallable} object that reduces the overhead of
calling compiled C functions, such as those generated using \texttt{numba} 
or Cython, directly from Python."
?

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -877,30 +873,28 @@ \subsubsection*{Sparse matrices}
 
 SciPy has provided special functions and leveraged BLAS and
 LAPACK\cite{LAPACK} routines for many years. SciPy now additionally
-includes Cython\cite{behnel2011cython} wrappers for:
```
The colon is not necessary here according to rule 2 of:
http://www.kentlaw.edu/academics/lrw/grinker/LwtaColons.htm

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -877,30 +873,28 @@ \subsubsection*{Sparse matrices}
 SciPy has provided special functions and leveraged BLAS and
 LAPACK\cite{LAPACK} routines for many years. SciPy now additionally
-includes Cython\cite{behnel2011cython} wrappers for:
-many BLAS and LAPACK routines (added in 2015), and the special functions 
```
Comma would be needed if we are separating three or more items with commas, but we're not.

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -919,7 +913,7 @@ \subsection*{Cython bindings for BLAS, LAPACK, and special}
 written in Fortran.  Not only can these low-level wrappers be used without a
 Fortran compiler, they can also be used without having to handle all the
-different Fortran compiler ABI's and name mangling schemes.
+different Fortran compiler ABIs and name mangling schemes.
```
Some would argue against apostrophe here (not possessive).

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -950,10 +944,10 @@ \subsection*{Statistical distributions}
 distributions and a framework to implement additional distributions. Given an
 expression for a probability distribution function (pdf), the framework will
```
Is any of this new? Consider removal except for the new stuff at the end.

tylerjereddy replied:
See #97 for this discussion -- maybe do that there

tylerjereddy replied:
ping @chrisb83

chrisb83 replied:
I think the framework to handle distributions is worth mentioning, but let's discuss in 97

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -975,73 +968,54 @@ \subsection*{Polynomial interpolators}
 
 The SciPy test suite is orchestrated by a continuous integration matrix that
 includes posix and Windows (32/64-bit) platforms managed by Travis CI and
-appveyor, respectively. Our tests cover Python versions 2.7, 3.4, 3.5, 3.6, and
```
What is new?

tylerjereddy replied:
If you're not sure, then maybe the issue already opened -- #97 regarding the stats section -- is the place to discuss this first. I think @chrisb83 has opened the issue.

tylerjereddy replied:
All of it -- CI is relatively new

tylerjereddy replied:
See the timeline figure you made

mdhaber replied:
That's why I asked. The timeline shows Travis CI starting in 2013. The section is about
"key technical improvements made in the last three years". Since we started Travis CI more than three years before the release of SciPy 1.0 in late 2017, I thought it was worth checking about the others.

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -975,73 +968,54 @@ \subsection*{Polynomial interpolators}
+The results are based on
+Python 2.7 performance on the master branch of the project using NumPy 1.8.2
+and Cython versions 0.27.3, 0.21.1, and 0.18 (for improved backward
+compatibility). Only the $L^2$ (Euclidean) norm is shown here, and to improve
```
This still needs revision; I assume it will done when this plot is formatted to be included in the report.
L^2 norm is not shown here. Elapsed time of something (what, exactly?) is shown.
What data is in the tree? What are the arguments to `query` and what do they mean?

tylerjereddy replied:
I disagree -- the L2 norm is definitely shown in this plot. Is the number 2 highlighted in the left side panel. There are different distance metrics possible for cKDTree and L2 is the one chosen for these results and displayed on the plot.

For the arguments to query, may refer the reader to something more detailed in the SI or simply the asv benchmark code itself.

mdhaber replied:
Ah, I see that the L^2 norm was an option set in the tests. But I meant that the phrase "Only the $L^2$ (Euclidean) norm is shown here..." suggests that the lines on the plot represent the L^2 norm of something. This could confuse a reader unfamiliar with either k-d trees or `asv` benchmarking.

mdhaber replied:
I will leave this to another PR.

mdhaber commented on scipy-1.0/scipy-optimize.tex:
```diff
@@ -22,65 +22,8 @@ \subsubsection*{Linear Optimization}
-While most \texttt{minimize} development prior to SciPy 1.0 has focused on unconstrained minimization, 
-% methods \texttt{SLSQP} and \texttt{COBYLA}\footnote{We note that any equality constraint $c_{\text{eq}}(x) = 0$ can be expressed as two inequality constraints $c_{\text{eq}}(x) \leq 0$ and $-c_{\text{eq}}(x) \leq 0$} 
-\texttt{SLSQP} supports fully general nonlinear programming. However, it does not provide special treatment for sparsity, limiting its utility to small problems. Accordingly, ongoing development efforts for \texttt{minimize} are focused on the addition of nonlinear programming algorithms for large, sparse problems. 
+The \texttt{minimize} function unifies several methods of finding local minima of nonlinear optimization problems in a common interface. Four new methods for unconstrained optimization were added to \texttt{minimize} in recent versions of SciPy: \texttt{dogleg}, \texttt{trust-ncg}, \texttt{trust-exact}, and \texttt{trust-krylov}. All are trust-region methods that build a local model of the object ...
```
As discussed in #102

mdhaber commented on scipy-1.0/scipy-optimize.tex:
```diff
@@ -22,65 +22,8 @@ \subsubsection*{Linear Optimization}
-% As far as I understand how brute works, the run time does not depend on the number of extrema, as it only polishes the best one
-% , or if the objective function has many extrema. 
-To address this, the
-\texttt{basinhopping} \cite{Wales1997} function uses random perturbation of the decision variables in conjunction with local minimization. The acceptance of the perturbed coordinates is dictated by the Metropolis test typically used in Monte Carlo simulations. 
```
`brute` and `basinhopping` are not new.

tylerjereddy replied:
great, marking as resolved

mdhaber commented on scipy-1.0/paper.tex:
```diff
@@ -848,22 +848,18 @@ \section*{Key technical improvements}
 and column-compressed formats (CSR and CSC, respectively). 
 These offer fast major-axis indexing and fast matrix-vector multiplication,
-and are used heavily in scikit-learn.
+and are used heavily throughout SciPy and dependent packages.
```
Why only mention scikit-learn? 
Can we cite a fact about a different package without citation?

tylerjereddy replied:
The general statement you made is probably fine -- I'll just mark this as resolved with your adjustment

mdhaber commented:
@tylerjereddy In retrospect, I should have split this into multiple PRs. Let me know what changes to revert so we can discuss them separately.

- [ ] I think we can resolve the LowLevelCallble section changes here. I suggested a change that would bring back the explicit mention of code generated by `numba` and Cython.
- [ ] I didn't make any sweeping changes to the stats section; we're leaving that to #97. I hope the language adjustments are OK. I can explain the reason for each change in detail if it would help.
- [ ] I agree that the changes to the Benchmark suite section are hard to follow because I re-ordered this and the test suite section. If you are willing to consider them here, the main changes are 1) the removal of the complete command for generating Figure~\ref{fig:asvbench} and 2) the crash course on `asv` use. The rationale is that the other sections aren't so detailed, so these seemed out of place. Just let me know if I should revert, though.
- [ ] GitHub is not doing a great job at identifying the the changes to the Test Suite section; they're pretty minor. The biggest is changing 
"There are more than $13000$ unit tests in the test suite, which is written for usage with the pytest runner, and with 87 \% and 45 \% line coverages for Python and compiled code, respectively at the SciPy 1.0 release point (Figure~\ref{fig:coverage})" 
to 
"There are more than $13,000$
unit tests in the test suite, which uses the \texttt{pytest} framework, and coverage of SciPy's approximately $204,000$ lines of code
has held steady at 76.5\% for the last six months (Figure~\ref{fig:coverage})"
This new sentence was based on an old version of the text, though. I was reviewing from a hard copy that didn't include some of the most recent changes. Perhaps it should be:
"There are more than $13,000$ unit tests in the test suite, which uses the \texttt{pytest} framework, with
 87 \% and 45 \% line coverages for Python and compiled code, respectively, at the SciPy 1.0 release point (Figure~\ref{fig:coverage})"
Let me know if I need to revert.
- [ ] I revised the `minimize` section considerably, but we had discussed this text in #102 
- [ ] I removed mention of `brute` and `basinhopping` from Global Minimization. GitHub shows a lot of green and red because I also removed a lot of commented-out lines.

So perhaps we can resolve most changes here, but I'll move the changes to "Test Suite" and "Benchmarking" to a separate PR?

mdhaber commented:
Split into separate PRs
